[ Tool ] Add support for linux riscv64 architecture

vhaudiquet opened the pull request:
This adds support for buildind desktop linux applications for riscv64 on the flutter tool, as well as basic riscv64 support for the tool.

This is a first move towards fixing #99963. Along with https://github.com/flutter/flutter/pull/178712 that allows building an engine for `riscv64`, this PR allows me to build and run the flutter 'hello world' app on RISC-V hardware:

<img width="1920" height="1080" alt="Screenshot from 2025-11-18 08-58-03" src="https://github.com/user-attachments/assets/5e90bc7f-49a7-41a8-82b7-5a5d3247938a" />

This allows someone who built a flutter engine and exposed it on a webserver to run commands like:
```
export FLUTTER_STORAGE_BASE_URL=http://your-server-url
flutter
flutter doctor
cd example/hello_world && flutter build linux
```

This means that even if Google does not want to support Flutter on RISC-V officially (hosting the artifacts), community-managed servers could host the RISC-V flutter engines/artifacts. It also means that internally, companies using Flutter on RISC-V would just have to build their engine and setup such a server, and it allows them to use the upstream Flutter tool.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

reidbaker reviewed (approved):
Approved but I would want someone from the engine team to take a look before this lands.

reidbaker reviewed (approved):
Not sure where you are on the process of landing this feature but I would have expected to see a test that built for risc-v as part of this pr to ensure we dont regress on the ability to build.

reidbaker commented:
At the minimum I would expect that you could author a test that attempted to build the risc-v code. That code would be required before we can publish an RISC-V engine binary.

vhaudiquet committed:
[ Tool ] Add riscv64 tool/commands build tests

vhaudiquet commented:
I tried to understand the test infrastructure and add a test that is basically a copy of the one present on arm64, adapted for riscv64. Is that what you had in mind?

reidbaker reviewed (approved):
Thank you for adding a build test. This should still be approved by someone on the engine team but I no longer see any issues.

gaaclarke reviewed (approved):
LGTM.  @reidbaker the tests here assert that the tool will attempt to build riscv64 but we don't actually buildit as part of CI right?  We should consider adding that as a follow up.

vhaudiquet commented:
I also submitted #178712 to allow building the engine (either locally or in CI). I started a discussion there, that is still ongoing it seems. 

Having official tool support could still be useful even if you don't plan/want to build the engine on CI, as it would allow third-party actors to use the official tool with their third-party built engine using `export FLUTTER_STORAGE_BASE_URL` as described; this is why I submitted two different PRs.

bkonyi reviewed (approved):
LGTM! Let's figure out our plan for https://github.com/flutter/flutter/pull/178712 first and try and land this PRs together. I'll leave final approval to @jtmcdole.

robert-ancell commented on packages/flutter_tools/lib/src/commands/build_linux.dart:
```diff
@@ -80,8 +82,7 @@ class BuildLinuxCommand extends BuildSubCommand {
       throwToolExit('"build linux" only supported on Linux hosts.');
     }
     // Cross-building for x64 targets on arm64 hosts is not supported.
```
And is arm64 to/from riscv64 cross-compilation also allowed?

robert-ancell reviewed (approved):
Some small questions about the cross-compilation checks, but otherwise I can't see any issues.

robert-ancell reviewed (approved):
Some small questions about the cross-compilation checks, but otherwise I can't see any issues.

zanderso reviewed (changes requested):
Noting that the other PR indicates that this is still experimental, the ability to target riscv64 should be guarded by a feature flag to disable by default on any channel except main.

See
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/features.dart  and
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/config.dart

vhaudiquet commented:
I pushed an update. This makes the `linux-riscv64` target available to the tool behind a feature flag, except on a riscv64 host. I feel like it makes sense to keep it on a riscv64 host, as anyway otherwise you can't cross-compile from riscv64. 

However, maybe that was not the right place in the code? Disabling it at the argument-parsing level can make sure that users will not even see the feature, but maybe this is not what we want.

Please review and tell me, I can change that, this is just a first draft for the feature. Thank you!

EDIT: It seems that this broke multiple tests. Fixing....

zanderso commented on packages/flutter_tools/lib/src/commands/build_linux.dart:
My suggestion would be to parse the argument, then if riscv64 is selected but the feature has not been enabled, you can emit a warning explaining that the feature is experimental and how to enable it.

vhaudiquet replied:
Ok, that seems more reasonable indeed. Thanks!

vhaudiquet replied:
Solved with last change

vhaudiquet committed:
[ Tool ] Support riscv64 architecture

This adds support for buildind desktop linux applications for riscv64 on the flutter tool, as well as basic riscv64 support for the tool.

vhaudiquet commented:
The google testing is the only one failing now. Is that a real fail, or a temporary/infra thing?

zanderso reviewed (approved):
lgtm

A Googler will need to follow step 5 of https://github.com/flutter/flutter/blob/master/docs/contributing/Feature-flags.md. @bkonyi would you be able to help out with that 2 line change?

jtmcdole committed:
Merge branch 'master' into riscv64-tool

jtmcdole commented:
This PR is 700+ commits behind master; updating to pick up changes.

jtmcdole commented:
Applying another fix internally
